src/server/lock_manager/deadlock.rs

```diff
@@ -757,55 +757,51 @@ where
         }
 
         let inner = Rc::clone(&self.inner);
-        let mut s = stream.map_err(Error::Grpc).filter_map(move |item| {
-            if let Ok(mut req) = item {
-                // It's possible the leader changes after registering this handler.
-                let mut inner = inner.borrow_mut();
-                if inner.role != Role::Leader {
-                    ERROR_COUNTER_METRICS.not_leader.inc();
-                    return future::ready(Some(Err(Error::Other(box_err!("leader changed")))));
-                }
-                let WaitForEntry {
-                    txn,
-                    wait_for_txn,
-                    key_hash,
-                    ..
-                } = req.get_entry();
-                let detect_table = &mut inner.detect_table;
-                let res = match req.get_tp() {
-                    DeadlockRequestType::Detect => {
-                        if let Some(deadlock_key_hash) =
-                            detect_table.detect(txn.into(), wait_for_txn.into(), *key_hash)
-                        {
-                            let mut resp = DeadlockResponse::default();
-                            resp.set_entry(req.take_entry());
-                            resp.set_deadlock_key_hash(deadlock_key_hash);
-                            Some(Ok((resp, WriteFlags::default())))
-                        } else {
-                            None
-                        }
-                    }
-                    DeadlockRequestType::CleanUpWaitFor => {
-                        detect_table.clean_up_wait_for(txn.into(), wait_for_txn.into(), *key_hash);
-                        None
-                    }
-                    DeadlockRequestType::CleanUp => {
-                        detect_table.clean_up(txn.into());
+        let mut s = stream.map_err(Error::Grpc).try_filter_map(move |mut req| {
```

Comment:
I remembered there should be a `and_then` before the `filter_map`.. It's removed by https://github.com/tikv/tikv/pull/8597/files#diff-92a18657172bd93f2d8b17a1e93cdd5aR765.